src/main/java/io/temporal/internal/statemachines/ActivityStateMachine.java

```diff
@@ -0,0 +1,404 @@
+            EventType.EVENT_TYPE_ACTIVITY_TASK_COMPLETED,
+            State.COMPLETED,
+            ActivityStateMachine::notifyCompleted)
+        .add(
+            State.STARTED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_FAILED,
+            State.FAILED,
+            ActivityStateMachine::notifyFailed)
+        .add(
+            State.STARTED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_TIMED_OUT,
+            State.TIMED_OUT,
+            ActivityStateMachine::notifyTimedOut)
+        .add(
+            State.SCHEDULED_EVENT_RECORDED,
+            ExplicitEvent.CANCEL,
+            State.SCHEDULED_ACTIVITY_CANCEL_COMMAND_CREATED,
+            ActivityStateMachine::createRequestCancelActivityTaskCommand)
+        .add(
+            State.SCHEDULED_ACTIVITY_CANCEL_COMMAND_CREATED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_CANCEL_REQUESTED,
+            State.SCHEDULED_ACTIVITY_CANCEL_EVENT_RECORDED,
+            ActivityStateMachine::notifyCanceledIfTryCancel)
+        .add(
+            State.SCHEDULED_ACTIVITY_CANCEL_COMMAND_CREATED,
+            CommandType.COMMAND_TYPE_REQUEST_CANCEL_ACTIVITY_TASK,
+            State.SCHEDULED_ACTIVITY_CANCEL_COMMAND_CREATED)
+        /*
+        This state transition is not possible.
+        It looks like it is valid when an event, handling of which requests activity
+        cancellation, precedes EVENT_TYPE_ACTIVITY_TASK_STARTED event.
+        But as all code execution happens in the event loop the STARTED event is
+        applied to the sate machine (as it is done for all command events before
+        the event loop invocation) before the cancellation request.
+        .add(
+            State.SCHEDULED_ACTIVITY_CANCEL_COMMAND_CREATED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_STARTED,
+            State.STARTED_ACTIVITY_CANCEL_COMMAND_CREATED)
+             */
+        .add(
+            State.SCHEDULED_ACTIVITY_CANCEL_COMMAND_CREATED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_TIMED_OUT,
+            State.TIMED_OUT,
+            ActivityStateMachine::cancelCommandNotifyTimedOut)
+        .add(
+            State.SCHEDULED_ACTIVITY_CANCEL_EVENT_RECORDED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_CANCELED,
+            State.CANCELED,
+            ActivityStateMachine::notifyCanceled)
+        .add(
+            State.SCHEDULED_ACTIVITY_CANCEL_EVENT_RECORDED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_STARTED,
+            State.STARTED_ACTIVITY_CANCEL_EVENT_RECORDED)
+        .add(
+            State.STARTED_ACTIVITY_CANCEL_COMMAND_CREATED,
+            EventType.EVENT_TYPE_ACTIVITY_TASK_CANCEL_REQUESTED,
+            State.STARTED_ACTIVITY_CANCEL_EVENT_RECORDED,
+            ActivityStateMachine::notifyCanceledIfTryCancel)
+        .add(
+            State.STARTED_ACTIVITY_CANCEL_COMMAND_CREATED,
+            CommandType.COMMAND_TYPE_REQUEST_CANCEL_ACTIVITY_TASK,
+            State.STARTED_ACTIVITY_CANCEL_COMMAND_CREATED)
+        .add(
```

Comment:
can we group the transitions by state? You have a bunch of transitions from state.STARTED above...feels like you should move this cancel up there as well

Reply:
Done